### PR TITLE
Offset query support

### PR DESCRIPTION
**Pull request by wemrysi**
Adds support for an optional offset into the source data when querying, introduces the `Seek` loader for datasources that support seeking to a point in the resource to implement and the corresponding `UpsertSink` for destinations that reports offsets into the resource as they are processed.

[ch10872]

### CODE REVIEW

File: connector/src/main/scala/quasar/connector/datasource/BatchLoader.scala [hunk 1]
```diff
@@ -16,31 +16,45 @@
...
       case Full(g) => Full((f[A] _) compose g)
+      case Seek(g) => Seek((q, o) => f(g(q, o)))
     }
 }
 
 object BatchLoader {
   final case class Full[F[_], Q, A](load: Q => F[A])
       extends BatchLoader[F, Q, A]
+
+  final case class Seek[F[_], Q, A](load: (Q, Option[Offset]) => F[A])
```
Comment (alissapajer) [inline connector/src/main/scala/quasar/connector/datasource/BatchLoader.scala]: 🎉

File: impl/src/main/scala/quasar/impl/QuasarDatasource.scala [hunk 2]
```diff
@@ -14,75 +14,75 @@
...
 
 import slamdata.Predef._
+
 import quasar.api.datasource.DatasourceType
 import quasar.api.resource._
 import quasar.connector.datasource.Datasource
 import quasar.qscript.{InterpretedRead, QScriptEducated}
 
-import scalaz.~>
+import cats.~>
```
Comment (alissapajer) [inline impl/src/main/scala/quasar/impl/QuasarDatasource.scala]: 👍 to the spattering of `scalaz` => `cats` changes - I like the approach of switching to `cats` incrementally in places where it's easy to do so

**Comment by alissapajer** [comment]
build errored due to ch10816 - flaky test in `DefaultResultPushSpec`

I restarted the build

File: connector/src/main/scala/quasar/connector/DataEvent.scala [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import cats.instances.byte._
+import cats.syntax.eq._
+import cats.syntax.functor._
+import cats.syntax.show._
+
+import fs2.Chunk
+
+sealed trait DataEvent[+I, +O] extends Product with Serializable
+
+object DataEvent {
```
Comment (alissapajer) [inline connector/src/main/scala/quasar/connector/DataEvent.scala]: It's interesting to observe that records can be `Create`d without declaring what the ids (`I`s) are for those rows. This means that this code design requires the ids to be included as first-class data (i.e. as values in the `Chunk[Byte]`). We've discussed this before; I was just noticing how this requirement ended up being reified in the types. I don't have an argument at this point against this design; I just kept coming back to this thought during the PR review and so wanted to write it down :) Is this design limiting us in the future in any way?

File: connector/src/main/scala/quasar/connector/ResourceError.scala [hunk 4]
```diff
@@ -20,26 +20,78 @@ import slamdata.Predef._
...
+      detail: Option[String],
+      cause: Option[Throwable])
+      extends ResourceError
+
+  final case class NotAResource(path: ResourcePath) extends ResourceError {
+    val detail = None
+    val cause = None
+  }
+
+  final case class SeekUnspported(path: ResourcePath) extends ResourceError {
```
Comment (wemrysi) [inline connector/src/main/scala/quasar/connector/ResourceError.scala]: RIP spelling

File: connector/src/main/scala/quasar/connector/DataEvent.scala (hunk 3 above)
Comment (wemrysi) [reply connector/src/main/scala/quasar/connector/DataEvent.scala]: That is an interesting observation. I can't think of any limitations this is imposing atm, as it seems intrinsic to the problem of correlating many output rows with a single input row that the id be part of the data.

**Pull request merged**